### PR TITLE
Allow range(stop) with default start for override grammar

### PR DESCRIPTION
**Pull request by victorjoos**
## Motivation

To make the range function more like the default python range(), and easier to type, I added a check if only start was added to set stop to start and start to zero.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

tested by adding a parametrize argument in `test_basic_sweeper.py`.

## Related Issues and PRs

N/A

### CODE REVIEW

**Comment by Jasha10** [comment]
Thanks for the PR!

It would be good to have a few more tests -- see the `tests/test_overrides_parser.py` file, where most of the `range` tests currently reside.
In particular, I think it would be good to pin down behavior for the following:
- behavior when some inputs are zero or negative, e.g. range(-10, step=1), range(10, step=-1), range(0), ...
  for some of these cases, it is not immediately obvious to me what the behavior should be.
- behavior when called on float inputs

**Comment by victorjoos** [comment]
Aaah, Thanks for the tips !

I was a bit fast in implementing this, didn't notice I had to also modify the parser. Let me modify the right files.

Not sure about the float inputs, but for the other cases I would say the best is to compare to the default python range implementation: 

 - `range(-10)  == []`
 - `range(-2, step=-1) == [0, -1]`
 - `range(0) == []`
 - `range(10, step=-1) == []`

**Comment by victorjoos** [comment]
Actually didn't need to change the parser. I added some tests and a new example in the docs.

File: news/1664.feature [hunk 1]
```diff
@@ -0,0 +1 @@
+Allow range() in override grammar to have only one argument (the stop value).
```
Comment (jieru-hu) [inline news/1664.feature:1]: pls add a line after this.

**Review by jieru-hu — APPROVED** [review]
one nit, otherwise lgtm. Thanks!

**Comment by jieru-hu** [comment]
@omry  could you take a look as well?

**Comment by omry** [comment]
Of course. but as I am in the middle of releasing 1.1 this will wait.

File: tests/test_basic_sweeper.py [hunk 2]
```diff
@@ -46,6 +46,7 @@
             id="batches_of_2",
         ),
         param(["a=range(0,3)"], None, [[["a=0"], ["a=1"], ["a=2"]]], id="range"),
+        param(["a=range(3)"], None, [[["a=0"], ["a=1"], ["a=2"]]], id="range_no_start"),
```
Comment (omry) [inline tests/test_basic_sweeper.py:49]: This is testing the basic sweeper. remove this test, it's not needed here.

File: tests/test_overrides_parser.py [hunk 3]
```diff
@@ -402,6 +402,19 @@ def test_simple_choice_sweep(value: str, expected: Any) -> None:
...
+            "range(-10, step=-1)",
+            RangeSweep(start=0, stop=-10, step=-1),
+            id="no_start_negative",
+        ),
+        param("range(5.5)", RangeSweep(start=0, stop=5.5, step=1), id="no_start_float"),
+        param(
+            "range(5.5, step=0.5)",
+            RangeSweep(start=0, stop=5.5, step=0.5),
+            id="no_start_step_float",
+        ),
```
Comment (omry) [inline tests/test_overrides_parser.py:417]: Add a few tests to test_eval, it also checks the resulting output and not just the parsed object.

File: hydra/_internal/grammar/grammar_functions.py [hunk 4]
```diff
@@ -184,7 +184,9 @@ def choice(
 
 
 def range(
-    start: Union[int, float], stop: Union[int, float], step: Union[int, float] = 1
+    start: Union[int, float],
+    stop: Optional[Union[int, float]] = None,
+    step: Union[int, float] = 1,
```
Comment (omry) [inline hydra/_internal/grammar/grammar_functions.py:189]: Add a sentence to the function doc describing the new functionality.

File: news/1664.feature (hunk 1 above)
Comment (omry) [inline news/1664.feature:1]:
```suggestion
Allow range() in override grammar to have only one argument (the stop value), e.g. range(3)
```

**Review by omry — CHANGES_REQUESTED** [review]
Thanks. Looking pretty good.
See comments.

**Comment by victorjoos** [comment]
Thank you, sorry for the delay. I implemented your suggestions.

**Review by omry — APPROVED** [review]
Looking good, thanks.

**Comment by omry** [comment]
We have some existing CI failures we are looking at. Will likely need you to rebase once those are fixed to get CI green again.

**Comment by jieru-hu** [comment]
@victorjoos  could you rebase master and force push to this branch? Many thanks!

**Pull request merged**